rfcs/0174-improve-terminology.md

```diff
@@ -0,0 +1,234 @@
+[design]: #detailed-design
+
+## Build terminology
+
+There's a lot of terminology in the Nix build process that is still used, but isn't very helpful anymore. Let's look at the basic Nix build process:
+
+```mermaid
+graph LR;
+  expr[Nix expression that contains a derivation] -->|nix eval| drv[.drv file];
+  drv -->|nix instantiate| storedrv[Store path with .drv file];
+  storedrv -->|nix realise| output[Store path with output];
+``` 
+
+This is usually done implicitly by the `nix build` command. However, there's a lot more complexity at hand! Let's look at some of the caveats that can apply:
+
+- Nix expression doesn't necessarily contain a derivation. Using `nix eval` to evaluate it will produce a result without producing a .drv file
+- Nix can be configured to not keep .drv files in the store using a setting `keep-derivations` (default is `true`)
+- Fixed-Output Derivations, or FODs, are mostly not about derivations. They are about a type of store path that has a well-known hash that is compared with the actual hash
+- Import From Derivation, or IFD, isn't about derivations, either! It is about reading data from a store path
+- Packages that exist in Nixpkgs are often called "derivations", despite the existence of a buildspec in .drv file being an implementation detail
+- Nix store outputs that aren't exposed in Nixpkgs are called "derivations". Examples include store paths passed to `src`, symlink farms and other kinds of store paths used in implementation, but not exposed to the end user
+
+In summary, "derivation" concept is an implementation detail. The term, however, is applied to many user-facing scenarios. As a result, the term becomes overloaded, confusing and difficult to comprehend. We should treat derivations as an implementation detail.
+
+Continuing on the topic of derivations, it is clear that store path outputs aren't always directly produced by a derivation. It is not very useful to think in "derivation -> output" chains. Instead, we can use a term "installable" that is [already established](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix#installables) to simplify the terminology:
+
+- `nix build` is a command that converts an installable to an output store path or fails. It is a simple "input -> output" relationship
+- In this model, what is written to the store (aside from the output) are just *dependencies*. This includes plain dependencies (like software libraries), FODs, implementation-based store paths and .drv files
+- Installables are usually packages or runnable scripts. That makes them easy to conceptualize, and might give additional intuition into the future Flake transition. The current unstable schema for Flakes uses installables for all outputs. That said, Flakes also allow definining custom outputs that can be arbitrary
+
+With all that reasoning, here are the actions proposed:
+
+- Consider renaming IFD (Import From Derivation). Here are possible options:
+  - Import From Store Path (IFSP)
```

Comment:
Isn't this ambiguous? `import (builtins.fetchGit ...)` is also an import from a store-path, but not IFD.

Reply:
Are you sure it's not IFD? Our current definition for IFD suggests that it should be:

> Passing an expression expr that evaluates to a [store path](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) to any built-in function which reads from the filesystem constitutes Import From Derivation (IFD):
>
> - [import](https://nixos.org/manual/nix/stable/language/builtins#builtins-import) expr

If the example you gave is indeed not IFD, we should rethink not only our term used, but also the definition.

Reply:
Minimal reproducer:

```
$ nix-instantiate --no-allow-import-from-derivation -E 'let git = builtins.fetchGit ./Projects/nix; in [ "${git}" (import "${git}")._type ]' --strict --eval
[ "/nix/store/xrkxv25fwqs241c5w2xvjaw7mrhs8daq-source" "flake" ]
```

Below `./Projects/nix` is a git checkout from Nix.

IIUC the error is indeed only thrown for _drvs_ being in string context of an import: https://github.com/NixOS/nix/blob/12ec315/src/libexpr/primops.cc#L74-L81